C Compiler error when using  ?fn (...)

### Describe the bug

Building the `get-started.v` program described below causes a C compiler build error.


### Reproduction Steps

- Clone https://github.com/mike-ward/gui/tree/c-bug-event
- cd into `examples` folder
- `v -path "/Users/mike/gui/src|@vlib|@vmodules" run get-started.v`

(substitute the `/Users/mike/gui/src` with cloned source

### Expected Behavior

Builds program

### Current Behavior

```
~~~~~~~~~~~~~~~~~~~
cc: /tmp/v_501/get-started.01JT8RAFQ1DA0FANRBEKKMJQ4S.tmp.c:20788:482: warning: implicit conversion from enumeration type 'sokol__gfx__LoadAction' to different enumeration type 'sg_load_action' (aka 'enum sg_load_action') [-Wenum-conversion]
...
cc: 26 warnings and 1 error generated.
(note: the original output was 103 lines long; it was truncated to its first 12 lines + the last line)
================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error:
==================
```

### Possible Solution

Line 38 in  `/src/gui/view_button.v` is where the problem originates. 

Changing from `fn(...)` to `?fn(...)` revealed the issue. (also lines 95, 103). 

### Additional Information/Context

_No response_

### V version

V 0.4.10 496c017

### Environment details (OS name and version, etc.)

|V full version      |V 0.4.10 da3112e.496c017
|:-------------------|:-------------------
|OS                  |macos, macOS, 15.4.1, 24E263
|Processor           |8 cpus, 64bit, little endian, Apple M2
|Memory              |0.15GB/8GB
|                    |
|V executable        |/Users/mike/Documents/github/v/v
|V last modified time|2025-05-01 12:03:30
|                    |
|V home dir          |OK, value: /Users/mike/Documents/github/v
|VMODULES            |OK, value: /Users/mike/.vmodules
|VTMP                |OK, value: /tmp/v_501
|Current working dir |OK, value: /Users/mike/Documents/github/gui/src/examples
|                    |
|env VFLAGS          |"-message-limit 5 -path /Users/mike/Documents/github/gui/src|@vlib|@vmodules"
|                    |
|Git version         |git version 2.49.0
|V git status        |weekly.2025.17-35-g496c0171
|.git/config present |true
|                    |
|cc version          |Apple clang version 17.0.0 (clang-1700.0.13.3)
|gcc version         |Apple clang version 17.0.0 (clang-1700.0.13.3)
|clang version       |Apple clang version 17.0.0 (clang-1700.0.13.3)
|tcc version         |tcc version 0.9.28rc 2024-02-05 HEAD@105d70f7 (AArch64 Darwin)
|tcc git status      |thirdparty-macos-arm64 c8df4e27
|emcc version        |N/A
|glibc version       |N/A
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

Odd, my last push didn't update. I guess I should have committed before renaming the branch. Branch updated now. Here's the full output from the updated branch.
```
╰─ ▶ v -show-c-output run get-started.v
======== Output of the C Compiler (cc) ========
In file included from /tmp/v_501/get-started.01JT9404FZG01DAZNQCAD6NHP4.tmp.c:2018:
/Users/mike/Documents/github/v/thirdparty/sokol/sokol_app.h:4106:72: warning: 'NSBorderlessWindowMask' is deprecated: first deprecated in macOS 10.12 [-Wdeprecated-declarations]
 4106 |                                                              styleMask:NSBorderlessWindowMask
      |                                                                        ^~~~~~~~~~~~~~~~~~~~~~
      |                                                                        NSWindowStyleMaskBorderless
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/AppKit.framework/Headers/NSWindow.h:1043:32: note: 'NSBorderlessWindowMask' has been explicitly marked deprecated here
 1043 | static const NSWindowStyleMask NSBorderlessWindowMask API_DEPRECATED_WITH_REPLACEMENT("NSWindowStyleMaskBorderless", macos(10.0,10.12)) = NSWindowStyleMaskBorderless;
      |                                ^
In file included from /tmp/v_501/get-started.01JT9404FZG01DAZNQCAD6NHP4.tmp.c:6269:
/Users/mike/Documents/github/v/vlib/clipboard/clipboard_darwin.m:7:66: warning: 'NSStringPboardType' is deprecated: first deprecated in macOS 10.14 [-Wdeprecated-declarations]
    7 |         NSString *ns_clip = [((__bridge NSPasteboard*)pb) stringForType:NSStringPboardType]; //NSPasteboardTypeString
      |                                                                         ^~~~~~~~~~~~~~~~~~
      |                                                                         NSPasteboardTypeString
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/AppKit.framework/Headers/NSPasteboard.h:432:32: note: 'NSStringPboardType' has been explicitly marked deprecated here
  432 | APPKIT_EXTERN NSPasteboardType NSStringPboardType API_DEPRECATED_WITH_REPLACEMENT("NSPasteboardTypeString", macos(10.0,10.14));
      |                                ^
In file included from /tmp/v_501/get-started.01JT9404FZG01DAZNQCAD6NHP4.tmp.c:6269:
/Users/mike/Documents/github/v/vlib/clipboard/clipboard_darwin.m:11:9: warning: returning 'NS_RETURNS_INNER_POINTER const char *' from a function with result type 'char *' discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]
   11 |         return [ns_clip UTF8String];
      |                ^~~~~~~~~~~~~~~~~~~~
/Users/mike/Documents/github/v/vlib/clipboard/clipboard_darwin.m:17:44: warning: 'NSStringPboardType' is deprecated: first deprecated in macOS 10.14 [-Wdeprecated-declarations]
   17 |         [pb declareTypes:[NSArray arrayWithObject:NSStringPboardType] owner:nil];
      |                                                   ^~~~~~~~~~~~~~~~~~
      |                                                   NSPasteboardTypeString
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/AppKit.framework/Headers/NSPasteboard.h:432:32: note: 'NSStringPboardType' has been explicitly marked deprecated here
  432 | APPKIT_EXTERN NSPasteboardType NSStringPboardType API_DEPRECATED_WITH_REPLACEMENT("NSPasteboardTypeString", macos(10.0,10.14));
      |                                ^
In file included from /tmp/v_501/get-started.01JT9404FZG01DAZNQCAD6NHP4.tmp.c:6269:
/Users/mike/Documents/github/v/vlib/clipboard/clipboard_darwin.m:18:43: warning: 'NSStringPboardType' is deprecated: first deprecated in macOS 10.14 [-Wdeprecated-declarations]
   18 |         bool ret = [pb setString:ns_clip forType:NSStringPboardType];
      |                                                  ^~~~~~~~~~~~~~~~~~
      |                                                  NSPasteboardTypeString
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/AppKit.framework/Headers/NSPasteboard.h:432:32: note: 'NSStringPboardType' has been explicitly marked deprecated here
  432 | APPKIT_EXTERN NSPasteboardType NSStringPboardType API_DEPRECATED_WITH_REPLACEMENT("NSPasteboardTypeString", macos(10.0,10.14));
      |                                ^
/tmp/v_501/get-started.01JT9404FZG01DAZNQCAD6NHP4.tmp.c:13882:6: warning: initializing 'u8 *' (aka 'unsigned char *') with an expression of type 'char *' converts between pointers to integer types where one is of the unique plain 'char' type and the other is not [-Wpointer-sign]
 13882 |         u8* utf8_clip = darwin_get_pasteboard_text(cb->pb);
       |             ^           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/v_501/get-started.01JT9404FZG01DAZNQCAD6NHP4.tmp.c:14808:103: warning: implicit conversion from enumeration type 'sokol__gfx__LoadAction' to different enumeration type 'sg_load_action' (aka 'enum sg_load_action') [-Wenum-conversion]
 14808 |         sokol__gfx__ColorAttachmentAction color_action = ((sokol__gfx__ColorAttachmentAction){.load_action = sokol__gfx__LoadAction__clear,.store_action = 0,.clear_value = ((sokol__gfx__Color){.r = r,.g = g,.b = b,.a = a,}),});
       |                                                                                              ~               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/v_501/get-started.01JT9404FZG01DAZNQCAD6NHP4.tmp.c:15234:32: warning: implicit conversion from enumeration type 'sokol__gfx__PixelFormat' to different enumeration type 'sg_pixel_format' (aka 'enum sg_pixel_format') [-Wenum-conversion]
 15234 |         env.defaults.color_format = (*(sokol__gfx__PixelFormat*)_t1.data);
       |                                   ~  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/v_501/get-started.01JT9404FZG01DAZNQCAD6NHP4.tmp.c:15241:32: warning: implicit conversion from enumeration type 'sokol__gfx__PixelFormat' to different enumeration type 'sg_pixel_format' (aka 'enum sg_pixel_format') [-Wenum-conversion]
 15241 |         env.defaults.depth_format = (*(sokol__gfx__PixelFormat*)_t2.data);
       |                                   ~  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/v_501/get-started.01JT9404FZG01DAZNQCAD6NHP4.tmp.c:15262:29: warning: implicit conversion from enumeration type 'sokol__gfx__PixelFormat' to different enumeration type 'sg_pixel_format' (aka 'enum sg_pixel_format') [-Wenum-conversion]
 15262 |         swapchain.color_format = (*(sokol__gfx__PixelFormat*)_t1.data);
       |                                ~  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/v_501/get-started.01JT9404FZG01DAZNQCAD6NHP4.tmp.c:15269:29: warning: implicit conversion from enumeration type 'sokol__gfx__PixelFormat' to different enumeration type 'sg_pixel_format' (aka 'enum sg_pixel_format') [-Wenum-conversion]
 15269 |         swapchain.depth_format = (*(sokol__gfx__PixelFormat*)_t2.data);
       |                                ~  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/v_501/get-started.01JT9404FZG01DAZNQCAD6NHP4.tmp.c:15301:24: warning: implicit conversion from enumeration type 'sokol__sapp__MouseCursor' to different enumeration type 'sapp_mouse_cursor' (aka 'enum sapp_mouse_cursor') [-Wenum-conversion]
 15301 |         sapp_set_mouse_cursor(cursor);
       |         ~~~~~~~~~~~~~~~~~~~~~ ^~~~~~
/tmp/v_501/get-started.01JT9404FZG01DAZNQCAD6NHP4.tmp.c:15338:172: warning: implicit conversion from enumeration type 'sapp_event_type' (aka 'enum sapp_event_type') to different enumeration type 'sokol__sapp__EventType' [-Wenum-conversion]
 15338 |         return str_intp(3, _MOV((StrIntpData[]){{_SLIT("evt: frame_count="), 0xfe08, {.d_u64 = e->frame_count}}, {_SLIT(", type="), 0xfe10, {.d_s = sokol__sapp__EventType_str(e->type)}}, {_SLIT0, 0, { .d_c = 0 }}}));
       |                                                                                                                                                     ~~~~~~~~~~~~~~~~~~~~~~~~~~ ~~~^~~~
/tmp/v_501/get-started.01JT9404FZG01DAZNQCAD6NHP4.tmp.c:15878:216: warning: implicit conversion from enumeration type 'sokol__gfx__BlendFactor' to different enumeration type 'sg_blend_factor' (aka 'enum sg_blend_factor') [-Wenum-conversion]
 15878 |         alpha_pipdesc.colors[0] = ((sokol__gfx__ColorTargetState){.pixel_format = 0,.write_mask = 0,.blend = ((sokol__gfx__BlendState){.enabled = true,.src_factor_rgb = sokol__gfx__BlendFactor__src_alpha,.dst_factor_rgb = sokol__gfx__BlendFactor__one_minus_src_alpha,.op_rgb = 0,.src_factor_alpha = 0,.dst_factor_alpha = 0,.op_alpha = 0,}),});
       |                                                                                                                                       ~                                                                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/v_501/get-started.01JT9404FZG01DAZNQCAD6NHP4.tmp.c:15878:163: warning: implicit conversion from enumeration type 'sokol__gfx__BlendFactor' to different enumeration type 'sg_blend_factor' (aka 'enum sg_blend_factor') [-Wenum-conversion]
 15878 |         alpha_pipdesc.colors[0] = ((sokol__gfx__ColorTargetState){.pixel_format = 0,.write_mask = 0,.blend = ((sokol__gfx__BlendState){.enabled = true,.src_factor_rgb = sokol__gfx__BlendFactor__src_alpha,.dst_factor_rgb = sokol__gfx__BlendFactor__one_minus_src_alpha,.op_rgb = 0,.src_factor_alpha = 0,.dst_factor_alpha = 0,.op_alpha = 0,}),});
       |                                                                                                                                       ~                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/v_501/get-started.01JT9404FZG01DAZNQCAD6NHP4.tmp.c:15883:214: warning: implicit conversion from enumeration type 'sokol__gfx__BlendFactor' to different enumeration type 'sg_blend_factor' (aka 'enum sg_blend_factor') [-Wenum-conversion]
 15883 |         add_pipdesc.colors[0] = ((sokol__gfx__ColorTargetState){.pixel_format = 0,.write_mask = 0,.blend = ((sokol__gfx__BlendState){.enabled = true,.src_factor_rgb = sokol__gfx__BlendFactor__src_alpha,.dst_factor_rgb = sokol__gfx__BlendFactor__one,.op_rgb = 0,.src_factor_alpha = 0,.dst_factor_alpha = 0,.op_alpha = 0,}),});
       |                                                                                                                                     ~                                                                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/v_501/get-started.01JT9404FZG01DAZNQCAD6NHP4.tmp.c:15883:161: warning: implicit conversion from enumeration type 'sokol__gfx__BlendFactor' to different enumeration type 'sg_blend_factor' (aka 'enum sg_blend_factor') [-Wenum-conversion]
 15883 |         add_pipdesc.colors[0] = ((sokol__gfx__ColorTargetState){.pixel_format = 0,.write_mask = 0,.blend = ((sokol__gfx__BlendState){.enabled = true,.src_factor_rgb = sokol__gfx__BlendFactor__src_alpha,.dst_factor_rgb = sokol__gfx__BlendFactor__one,.op_rgb = 0,.src_factor_alpha = 0,.dst_factor_alpha = 0,.op_alpha = 0,}),});
       |                                                                                                                                     ~                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/v_501/get-started.01JT9404FZG01DAZNQCAD6NHP4.tmp.c:16305:217: warning: implicit conversion from enumeration type 'sokol__gfx__Wrap' to different enumeration type 'sg_wrap' (aka 'enum sg_wrap') [-Wenum-conversion]
 16305 |         sokol__gfx__SamplerDesc smp_desc = ((sokol__gfx__SamplerDesc){.min_filter = sokol__gfx__Filter__linear,.mag_filter = sokol__gfx__Filter__linear,.mipmap_filter = 0,.wrap_u = sokol__gfx__Wrap__clamp_to_edge,.wrap_v = sokol__gfx__Wrap__clamp_to_edge,.wrap_w = 0,.min_lod = 0,.max_lod = 0,.border_color = 0,.compare = 0,.max_anisotropy = 0,.label = 0,.gl_sampler = 0,.mtl_sampler = 0,.d3d11_sampler = 0,.wgpu_sampler = 0,});
       |                                                                      ~                                                                                                                                                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/v_501/get-started.01JT9404FZG01DAZNQCAD6NHP4.tmp.c:16305:175: warning: implicit conversion from enumeration type 'sokol__gfx__Wrap' to different enumeration type 'sg_wrap' (aka 'enum sg_wrap') [-Wenum-conversion]
 16305 |         sokol__gfx__SamplerDesc smp_desc = ((sokol__gfx__SamplerDesc){.min_filter = sokol__gfx__Filter__linear,.mag_filter = sokol__gfx__Filter__linear,.mipmap_filter = 0,.wrap_u = sokol__gfx__Wrap__clamp_to_edge,.wrap_v = sokol__gfx__Wrap__clamp_to_edge,.wrap_w = 0,.min_lod = 0,.max_lod = 0,.border_color = 0,.compare = 0,.max_anisotropy = 0,.label = 0,.gl_sampler = 0,.mtl_sampler = 0,.d3d11_sampler = 0,.wgpu_sampler = 0,});
       |                                                                      ~                                                                                                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/v_501/get-started.01JT9404FZG01DAZNQCAD6NHP4.tmp.c:16305:119: warning: implicit conversion from enumeration type 'sokol__gfx__Filter' to different enumeration type 'sg_filter' (aka 'enum sg_filter') [-Wenum-conversion]
 16305 |         sokol__gfx__SamplerDesc smp_desc = ((sokol__gfx__SamplerDesc){.min_filter = sokol__gfx__Filter__linear,.mag_filter = sokol__gfx__Filter__linear,.mipmap_filter = 0,.wrap_u = sokol__gfx__Wrap__clamp_to_edge,.wrap_v = sokol__gfx__Wrap__clamp_to_edge,.wrap_w = 0,.min_lod = 0,.max_lod = 0,.border_color = 0,.compare = 0,.max_anisotropy = 0,.label = 0,.gl_sampler = 0,.mtl_sampler = 0,.d3d11_sampler = 0,.wgpu_sampler = 0,});
       |                                                                      ~                                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/v_501/get-started.01JT9404FZG01DAZNQCAD6NHP4.tmp.c:16305:78: warning: implicit conversion from enumeration type 'sokol__gfx__Filter' to different enumeration type 'sg_filter' (aka 'enum sg_filter') [-Wenum-conversion]
 16305 |         sokol__gfx__SamplerDesc smp_desc = ((sokol__gfx__SamplerDesc){.min_filter = sokol__gfx__Filter__linear,.mag_filter = sokol__gfx__Filter__linear,.mipmap_filter = 0,.wrap_u = sokol__gfx__Wrap__clamp_to_edge,.wrap_v = sokol__gfx__Wrap__clamp_to_edge,.wrap_w = 0,.min_lod = 0,.max_lod = 0,.border_color = 0,.compare = 0,.max_anisotropy = 0,.label = 0,.gl_sampler = 0,.mtl_sampler = 0,.d3d11_sampler = 0,.wgpu_sampler = 0,});
       |                                                                      ~              ^~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/v_501/get-started.01JT9404FZG01DAZNQCAD6NHP4.tmp.c:19220:15: error: initializing 'byte' (aka 'unsigned char') with an expression of incompatible type '_option_anon_fn_gui__buttoncfg_mut_gui__event_mut_gui__window' (aka 'struct _option_anon_fn_gui__buttoncfg_mut_gui__event_mut_gui__window')
 19220 |                 .on_click = (*(cfg)).on_click,
       |                             ^~~~~~~~~~~~~~~~~
/tmp/v_501/get-started.01JT9404FZG01DAZNQCAD6NHP4.tmp.c:369:59: note: expanded from macro 'HEAP'
  369 | #define HEAP(type, expr) ((type*)memdup((void*)&((type[]){expr}[0]), sizeof(type)))
      |                                                           ^~~~
/tmp/v_501/get-started.01JT9404FZG01DAZNQCAD6NHP4.tmp.c:20412:10116: warning: implicit conversion from enumeration type 'enum sapp_mouse_cursor' to different enumeration type 'sokol__sapp__MouseCursor' [-Wenum-conversion]
/tmp/v_501/get-started.01JT9404FZG01DAZNQCAD6NHP4.tmp.c:20788:664: warning: implicit conversion from enumeration type 'sokol__gfx__LoadAction' to different enumeration type 'sg_load_action' (aka 'enum sg_load_action') [-Wenum-conversion]
 20788 |         _const_gg__dontcare_pass = ((sokol__gfx__PassAction){.colors = {((sokol__gfx__ColorAttachmentAction){.load_action = sokol__gfx__LoadAction__dontcare,.store_action = 0,.clear_value = ((sokol__gfx__Color){.r = 1.0,.g = 1.0,.b = 1.0,.a = 1.0,}),}), ((sokol__gfx__ColorAttachmentAction){.load_action = sokol__gfx__LoadAction__dontcare,.store_action = 0,.clear_value = ((sokol__gfx__Color){.r = 1.0,.g = 1.0,.b = 1.0,.a = 1.0,}),}), ((sokol__gfx__ColorAttachmentAction){.load_action = sokol__gfx__LoadAction__dontcare,.store_action = 0,.clear_value = ((sokol__gfx__Color){.r = 1.0,.g = 1.0,.b = 1.0,.a = 1.0,}),}), ((sokol__gfx__ColorAttachmentAction){.load_action = sokol__gfx__LoadAction__dontcare,.store_action = 0,.clear_value = ((sokol__gfx__Color){.r = 1.0,.g = 1.0,.b = 1.0,.a = 1.0,}),})},.depth = {0},.stencil = {0},});
       |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               ~               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/v_501/get-started.01JT9404FZG01DAZNQCAD6NHP4.tmp.c:20788:482: warning: implicit conversion from enumeration type 'sokol__gfx__LoadAction' to different enumeration type 'sg_load_action' (aka 'enum sg_load_action') [-Wenum-conversion]
 20788 |         _const_gg__dontcare_pass = ((sokol__gfx__PassAction){.colors = {((sokol__gfx__ColorAttachmentAction){.load_action = sokol__gfx__LoadAction__dontcare,.store_action = 0,.clear_value = ((sokol__gfx__Color){.r = 1.0,.g = 1.0,.b = 1.0,.a = 1.0,}),}), ((sokol__gfx__ColorAttachmentAction){.load_action = sokol__gfx__LoadAction__dontcare,.store_action = 0,.clear_value = ((sokol__gfx__Color){.r = 1.0,.g = 1.0,.b = 1.0,.a = 1.0,}),}), ((sokol__gfx__ColorAttachmentAction){.load_action = sokol__gfx__LoadAction__dontcare,.store_action = 0,.clear_value = ((sokol__gfx__Color){.r = 1.0,.g = 1.0,.b = 1.0,.a = 1.0,}),}), ((sokol__gfx__ColorAttachmentAction){.load_action = sokol__gfx__LoadAction__dontcare,.store_action = 0,.clear_value = ((sokol__gfx__Color){.r = 1.0,.g = 1.0,.b = 1.0,.a = 1.0,}),})},.depth = {0},.stencil = {0},});
       |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         ~               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/v_501/get-started.01JT9404FZG01DAZNQCAD6NHP4.tmp.c:20788:300: warning: implicit conversion from enumeration type 'sokol__gfx__LoadAction' to different enumeration type 'sg_load_action' (aka 'enum sg_load_action') [-Wenum-conversion]
 20788 |         _const_gg__dontcare_pass = ((sokol__gfx__PassAction){.colors = {((sokol__gfx__ColorAttachmentAction){.load_action = sokol__gfx__LoadAction__dontcare,.store_action = 0,.clear_value = ((sokol__gfx__Color){.r = 1.0,.g = 1.0,.b = 1.0,.a = 1.0,}),}), ((sokol__gfx__ColorAttachmentAction){.load_action = sokol__gfx__LoadAction__dontcare,.store_action = 0,.clear_value = ((sokol__gfx__Color){.r = 1.0,.g = 1.0,.b = 1.0,.a = 1.0,}),}), ((sokol__gfx__ColorAttachmentAction){.load_action = sokol__gfx__LoadAction__dontcare,.store_action = 0,.clear_value = ((sokol__gfx__Color){.r = 1.0,.g = 1.0,.b = 1.0,.a = 1.0,}),}), ((sokol__gfx__ColorAttachmentAction){.load_action = sokol__gfx__LoadAction__dontcare,.store_action = 0,.clear_value = ((sokol__gfx__Color){.r = 1.0,.g = 1.0,.b = 1.0,.a = 1.0,}),})},.depth = {0},.stencil = {0},});
       |                                                                                                                                                                                                                                                                                                   ~               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/v_501/get-started.01JT9404FZG01DAZNQCAD6NHP4.tmp.c:20788:118: warning: implicit conversion from enumeration type 'sokol__gfx__LoadAction' to different enumeration type 'sg_load_action' (aka 'enum sg_load_action') [-Wenum-conversion]
 20788 |         _const_gg__dontcare_pass = ((sokol__gfx__PassAction){.colors = {((sokol__gfx__ColorAttachmentAction){.load_action = sokol__gfx__LoadAction__dontcare,.store_action = 0,.clear_value = ((sokol__gfx__Color){.r = 1.0,.g = 1.0,.b = 1.0,.a = 1.0,}),}), ((sokol__gfx__ColorAttachmentAction){.load_action = sokol__gfx__LoadAction__dontcare,.store_action = 0,.clear_value = ((sokol__gfx__Color){.r = 1.0,.g = 1.0,.b = 1.0,.a = 1.0,}),}), ((sokol__gfx__ColorAttachmentAction){.load_action = sokol__gfx__LoadAction__dontcare,.store_action = 0,.clear_value = ((sokol__gfx__Color){.r = 1.0,.g = 1.0,.b = 1.0,.a = 1.0,}),}), ((sokol__gfx__ColorAttachmentAction){.load_action = sokol__gfx__LoadAction__dontcare,.store_action = 0,.clear_value = ((sokol__gfx__Color){.r = 1.0,.g = 1.0,.b = 1.0,.a = 1.0,}),})},.depth = {0},.stencil = {0},});
       |                                                                                                             ~               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
26 warnings and 1 error generated.
===============================================
No such file or directory; code: 2
```

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -2295,6 +2295,7 @@ fn (mut g Gen) expr_with_tmp_var(expr ast.Expr, expr_typ ast.Type, ret_typ ast.T
 			g.writeln('${g.styp(ret_typ)} ${tmp_var};')
 		}
 		mut expr_is_fixed_array_var := false
+		mut fn_option_clone := false
 		if ret_typ_is_option {
 			if expr_typ_is_option && expr in [ast.StructInit, ast.ArrayInit, ast.MapInit] {
 				simple_assign = expr is ast.StructInit
@@ -2328,6 +2329,7 @@ fn (mut g Gen) expr_with_tmp_var(expr ast.Expr, expr_typ ast.Type, ret_typ ast.T
 							g.write('(')
 							g.write_fn_ptr_decl(&final_ret_sym.info, '')
 							g.write(')')
+							fn_option_clone = expr is ast.SelectorExpr && expr_typ.has_flag(.option)
 						}
 					}
 				}
@@ -2339,6 +2341,9 @@ fn (mut g Gen) expr_with_tmp_var(expr ast.Expr, expr_typ ast.Type, ret_typ ast.T
 			g.write('_result_ok(&(${styp}[]) { ')
 		}
 		g.expr_with_cast(expr, expr_typ, ret_typ)
+		if fn_option_clone {
+			g.write('.data')
+		}
 		if ret_typ_is_option {
 			if simple_assign {
 				g.writeln(';')

--- a/vlib/v/gen/c/struct.v
+++ b/vlib/v/gen/c/struct.v
@@ -742,21 +742,21 @@ fn (mut g Gen) struct_init_field(sfield ast.StructInitField, language ast.Langua
 						g.fixed_array_var_init(tmp_var, false, field_unwrap_sym.info.elem_type,
 							field_unwrap_sym.info.size)
 					} else {
-						g.struct_init_field_default(field_unwrap_typ, sfield)
+						g.struct_init_field_default(field_unwrap_typ, sfield, field_unwrap_sym)
 					}
 				}
 				else {
-					g.struct_init_field_default(field_unwrap_typ, sfield)
+					g.struct_init_field_default(field_unwrap_typ, sfield, field_unwrap_sym)
 				}
 			}
 		} else {
-			g.struct_init_field_default(field_unwrap_typ, sfield)
+			g.struct_init_field_default(field_unwrap_typ, sfield, field_unwrap_sym)
 		}
 		g.inside_cast_in_heap = inside_cast_in_heap // restore value for further struct inits
 	}
 }
 
-fn (mut g Gen) struct_init_field_default(field_unwrap_typ ast.Type, sfield &ast.StructInitField) {
+fn (mut g Gen) struct_init_field_default(field_unwrap_typ ast.Type, sfield &ast.StructInitField, field_unwrap_sym ast.TypeSymbol) {
 	if field_unwrap_typ != ast.voidptr_type && field_unwrap_typ != ast.nil_type
 		&& (sfield.expected_type.is_ptr() && !sfield.expected_type.has_flag(.shared_f))
 		&& !sfield.expected_type.has_flag(.option) && !field_unwrap_typ.is_any_kind_of_pointer()
@@ -769,6 +769,9 @@ fn (mut g Gen) struct_init_field_default(field_unwrap_typ ast.Type, sfield &ast.
 		g.expr_with_opt(sfield.expr, field_unwrap_typ, sfield.expected_type)
 	} else if sfield.expr is ast.LambdaExpr && sfield.expected_type.has_flag(.option) {
 		g.expr_opt_with_cast(sfield.expr, field_unwrap_typ, sfield.expected_type)
+	} else if field_unwrap_sym.kind == .function && sfield.expected_type.has_flag(.option) {
+		tmp_out_var := g.new_tmp_var()
+		g.expr_with_tmp_var(sfield.expr, field_unwrap_typ, sfield.expected_type, tmp_out_var)
 	} else {
 		g.left_is_opt = true
 		g.expr_with_cast(sfield.expr, field_unwrap_typ, sfield.expected_type)

--- a/vlib/v/tests/options/option_fn_struct_init_test.v
+++ b/vlib/v/tests/options/option_fn_struct_init_test.v
@@ -1,0 +1,47 @@
+type Baz = Foo
+
+@[heap]
+struct Foo {
+	a  int    = 123
+	b  string = 'b'
+	cb ?fn (&Foo)
+}
+
+@[heap]
+struct Bar {
+	a  int    = 321
+	b  string = 'a'
+	cb ?fn (&Baz)
+}
+
+fn g(a Foo) Foo {
+	return a
+}
+
+fn t(a Bar) Foo {
+	return g(a: 123, b: 'foo', cb: a.cb)
+}
+
+fn test_main() {
+	a := Foo{
+		cb: none
+	}
+
+	b := Foo{
+		cb: fn (a &Foo) {
+		}
+	}
+
+	w := Bar{
+		cb: b.cb
+	}
+
+	z := Bar{
+		cb: a.cb
+	}
+
+	t(a: 1, cb: a.cb)
+	t(a: 2, cb: b.cb)
+	t(a: 3, cb: w.cb)
+	t(a: 4, cb: z.cb)
+}

